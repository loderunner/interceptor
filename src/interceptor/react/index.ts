export { Provider as InterceptProvider, useIntercepts } from './intercept'
