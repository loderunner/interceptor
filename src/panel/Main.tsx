import * as React from 'react'
import { useMemo } from 'react'

import InterceptView from './InterceptView'
import RequestView from './RequestDetails'
import { useSelection } from './selection'
import WelcomeView from './WelcomeView'

import type { Intercept, Request } from '@/interceptor'

interface Props {
  className?: string
}

const Main = ({ className }: Props) => {
  const { selection, selectionType } = useSelection()

  const view: React.ReactNode = useMemo(() => {
    switch (selectionType) {
      case 'null':
        return <WelcomeView />
      case 'intercept': {
        const inter: Intercept = selection as Intercept
        return <InterceptView interceptId={inter.id} />
      }
      case 'request': {
        const req: Request = selection as Request
        return <RequestView requestId={req.id} />
      }
    }
  }, [selection, selectionType])

  return (
    <main id="main" className={`min-h-screen ${className}`}>
      {view}
    </main>
  )
}

export default Main
