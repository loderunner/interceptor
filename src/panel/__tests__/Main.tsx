/* eslint-disable @typescript-eslint/ban-ts-comment */
import { cleanup, render } from '@testing-library/react'
import * as React from 'react'
import '@testing-library/jest-dom'

import Main from '../Main'
import * as selection from '../selection'

jest.mock('../selection')
const mockedSelection = selection as jest.Mocked<typeof selection>

describe('[Main]', () => {
  afterEach(() => {
    cleanup()
  })

  it('should match empty selection snapshot', () => {
    mockedSelection.useSelection.mockImplementation(() => [
      null,
      () => {},
      'null',
    ])
    const { container } = render(<Main />)
    expect(container).toMatchSnapshot()
  })

  it('should match intercept selection snapshot', () => {
    mockedSelection.useSelection.mockImplementation(() => [
      { pattern: 'helloworld', enabled: true },
      () => {},
      'intercept',
    ])
    const { container } = render(<Main />)
    expect(container).toMatchSnapshot()
  })

  it('should match request selection snapshot', () => {
    mockedSelection.useSelection.mockImplementation(() => [
      {
        headers: {},
        initialPriority: 'Medium',
        method: 'GET',
        referrerPolicy: 'same-origin',
        url: 'https://example.com',
      },
      () => {},
      'request',
    ])
    const { container } = render(<Main />)
    expect(container).toMatchSnapshot()
  })
})
