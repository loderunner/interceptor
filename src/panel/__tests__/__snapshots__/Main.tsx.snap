// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Main] should match empty selection snapshot 1`] = `
<div>
  <main
    class=""
    id="main"
  >
    <div
      class="container flex min-h-screen flex-col items-center justify-center bg-slate-50 "
      id="welcome-view"
    >
      <div
        class="container max-w-xs"
      >
        <svg
          class="h-auto w-full fill-indigo-700"
        />
      </div>
      <div
        class="text-4xl font-bold"
      >
        Interceptor
      </div>
      <button
        class="my-6 rounded bg-indigo-700 px-9 py-3 text-sm font-semibold text-white hover:opacity-75 active:opacity-100"
        title="Add intercept"
      >
        Add Intercept
      </button>
    </div>
  </main>
</div>
`;

exports[`[Main] should match intercept selection snapshot 1`] = `
<div>
  <main
    class=""
    id="main"
  >
    <div
      class="max-w-5xl mx-24 mt-8 px-8 pt-2 grid grid-cols-1 gap-6"
    >
      <label
        class="block"
      >
        <span
          class="text-3xl font-bold"
        >
          Intercept
        </span>
      </label>
      <label
        class="mt-1 inline-flex items-center"
      >
        <input
          checked=""
          type="checkbox"
        />
        <span
          class="ml-1 text-sm text-gray-700"
        >
          Enabled
        </span>
      </label>
      <label
        class="mt-1 block"
      >
        <span
          class="text-sm text-gray-700"
        >
          If URL matches
        </span>
        <div
          class="flex items-center space-x-2"
        >
          <input
            class="block mt-1 text-sm flex-auto"
            placeholder="example.com"
            type="text"
            value="example.com"
          />
          <label
            class="inline-flex items-center"
          >
            <input
              type="checkbox"
            />
            <span
              class="ml-1 text-sm text-gray-700"
            >
              Regexp
            </span>
          </label>
        </div>
      </label>
      <label
        class="mt-1 inline-flex items-center"
      >
        <input
          type="checkbox"
        />
        <span
          class="ml-1 text-sm text-gray-700"
        >
          Intercept response
        </span>
      </label>
    </div>
  </main>
</div>
`;

exports[`[Main] should match request selection snapshot 1`] = `
<div>
  <main
    class=""
    id="main"
  >
    <div
      class="max-w-5xl mx-24 my-8 px-8 pt-2"
    >
      <div
        class="flex space-x-4 mb-4"
      >
        <button
          class="text-base rounded py-2 px-4 font-medium border bg-blue-50"
          disabled=""
        >
          Headers
        </button>
        <button
          class="text-base rounded py-2 px-4 font-medium border hover:bg-gray-300 active:bg-gray-400"
        >
          Body
        </button>
      </div>
      <label
        class="mt-1 inline-flex items-center"
      >
        <input
          type="checkbox"
        />
        <span
          class="ml-1 text-sm text-gray-700"
        >
          Intercept response
        </span>
      </label>
      <div
        class="mt-8 "
      >
        <span
          class="text-3xl font-bold"
        >
          www.example.com
        </span>
        <div
          class="my-8 grid grid-cols-[10rem_1fr] gap-x-4 gap-y-1"
        >
          <span
            class="text-right font-medium text-gray-500 select-none"
          >
            URL
          </span>
          <span
            class="flex overflow-x-hidden space-x-1"
          >
            <span
              class="text-ellipsis overflow-x-hidden whitespace-nowrap"
            >
              https://www.example.com/complete/search
            </span>
          </span>
          <span
            class="text-right font-medium text-gray-500 select-none"
          >
            Method
          </span>
          <span
            class="flex overflow-x-hidden space-x-1"
          >
            <span
              class="text-ellipsis overflow-x-hidden whitespace-nowrap"
            >
              POST
            </span>
          </span>
          <span
            class="text-lg font-semibold col-span-2 mt-4 select-none"
          >
            Query
          </span>
          <div
            class="group flex justify-end items-center space-x-1"
          >
            <button
              class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700 opacity-0 group-hover:opacity-100"
            >
              <svg
                class="fill-white"
              />
            </button>
            <span
              class="font-medium text-right text-gray-500 select-none"
            >
              q
            </span>
          </div>
          <span
            class="flex overflow-x-hidden space-x-1"
          >
            <span
              class="text-ellipsis overflow-x-hidden whitespace-nowrap"
            >
              toto
            </span>
          </span>
          <div
            class="group flex justify-end items-center space-x-1"
          >
            <button
              class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700 opacity-0 group-hover:opacity-100"
            >
              <svg
                class="fill-white"
              />
            </button>
            <span
              class="font-medium text-right text-gray-500 select-none"
            >
              client
            </span>
          </div>
          <span
            class="flex overflow-x-hidden space-x-1"
          >
            <span
              class="text-ellipsis overflow-x-hidden whitespace-nowrap"
            >
              gws-wiz
            </span>
          </span>
          <div
            class="relative col-span-2"
          >
            <button
              class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
              style="margin-left: 0rem; width: calc(100% - 0rem); z-index: 0;"
            >
              <div
                class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
              />
              <div
                class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
              >
                <svg
                  class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                />
              </div>
            </button>
          </div>
          <span
            class="text-lg font-semibold col-span-2 mt-4 select-none"
          >
            Headers
          </span>
          <div
            class="group flex justify-end items-center space-x-1"
          >
            <button
              class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700 opacity-0 group-hover:opacity-100"
            >
              <svg
                class="fill-white"
              />
            </button>
            <span
              class="font-medium text-right text-gray-500 select-none"
            >
              Accept
            </span>
          </div>
          <span
            class="flex overflow-x-hidden space-x-1"
          >
            <span
              class="text-ellipsis overflow-x-hidden whitespace-nowrap"
            >
              */*
            </span>
          </span>
          <div
            class="group flex justify-end items-center space-x-1"
          >
            <button
              class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700 opacity-0 group-hover:opacity-100"
            >
              <svg
                class="fill-white"
              />
            </button>
            <span
              class="font-medium text-right text-gray-500 select-none"
            >
              Referer
            </span>
          </div>
          <span
            class="flex overflow-x-hidden space-x-1"
          >
            <span
              class="text-ellipsis overflow-x-hidden whitespace-nowrap"
            >
              https://www.google.com/
            </span>
          </span>
          <div
            class="group flex justify-end items-center space-x-1"
          >
            <button
              class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700 opacity-0 group-hover:opacity-100"
            >
              <svg
                class="fill-white"
              />
            </button>
            <span
              class="font-medium text-right text-gray-500 select-none"
            >
              User-Agent
            </span>
          </div>
          <span
            class="flex overflow-x-hidden space-x-1"
          >
            <span
              class="text-ellipsis overflow-x-hidden whitespace-nowrap"
            >
              Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36
            </span>
          </span>
          <div
            class="group flex justify-end items-center space-x-1"
          >
            <button
              class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700 opacity-0 group-hover:opacity-100"
            >
              <svg
                class="fill-white"
              />
            </button>
            <span
              class="font-medium text-right text-gray-500 select-none"
            >
              Content-Type
            </span>
          </div>
          <span
            class="flex overflow-x-hidden space-x-1"
          >
            <span
              class="text-ellipsis overflow-x-hidden whitespace-nowrap"
            >
              application/json
            </span>
          </span>
          <div
            class="relative col-span-2"
          >
            <button
              class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
              style="margin-left: 0rem; width: calc(100% - 0rem); z-index: 0;"
            >
              <div
                class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
              />
              <div
                class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
              >
                <svg
                  class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                />
              </div>
            </button>
          </div>
          <span
            class="text-lg font-semibold col-span-2 mt-4 select-none"
          >
            Cookies
          </span>
          <div
            class="group flex justify-end items-center space-x-1"
          >
            <span
              class="font-medium text-right text-gray-500 select-none"
            >
              CONSENT
            </span>
          </div>
          <span
            class="flex overflow-x-hidden space-x-1"
          >
            <span
              class="text-ellipsis overflow-x-hidden whitespace-nowrap"
            >
              YES
            </span>
          </span>
          <div
            class="group flex justify-end items-center space-x-1"
          >
            <span
              class="font-medium text-right text-gray-500 select-none"
            >
              DATE
            </span>
          </div>
          <span
            class="flex overflow-x-hidden space-x-1"
          >
            <span
              class="text-ellipsis overflow-x-hidden whitespace-nowrap"
            >
              2022-02-20-16
            </span>
          </span>
        </div>
      </div>
      <div
        class="flex flex-col justify-items-stretch mt-8 hidden"
      >
        <div
          class="relative -ml-4"
        >
          <button
            class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
            style="margin-left: 1rem; width: calc(100% - 1rem); z-index: 1;"
          >
            <div
              class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
            />
            <div
              class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
            >
              <svg
                class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
              />
            </div>
          </button>
          <div
            class="mt-1 flex items-center space-x-1"
            style="margin-left: 0rem;"
          >
            <button
              class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
            >
              <svg
                class="fill-white"
              />
            </button>
            <pre
              class="text-amber-700"
            >
              hello
              :
            </pre>
             
            <div
              class="flex-auto"
            >
              <pre
                class="whitespace-pre-wrap break-all text-sky-400"
              >
                "world"
              </pre>
            </div>
          </div>
          <button
            class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
            style="margin-left: 1rem; width: calc(100% - 1rem); z-index: 1;"
          >
            <div
              class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
            />
            <div
              class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
            >
              <svg
                class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
              />
            </div>
          </button>
          <div
            class="mt-1 flex items-center space-x-1"
            style="margin-left: 0rem;"
          >
            <button
              class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700 invisible"
            >
              <svg
                class="fill-white"
              />
            </button>
            <pre
              class="text-amber-700"
            >
              obj
              :
            </pre>
             
            <pre
              class="flex items-center space-x-0.5"
            >
              {
              <button
                class="mx-0.5 rounded-sm bg-slate-400 hover:bg-slate-500 active:bg-slate-600"
              >
                <svg />
              </button>
              }
            </pre>
          </div>
          <div
            class="relative "
          >
            <button
              class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
              style="margin-left: 2rem; width: calc(100% - 2rem); z-index: 2;"
            >
              <div
                class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
              />
              <div
                class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
              >
                <svg
                  class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                />
              </div>
            </button>
            <div
              class="mt-1 flex items-center space-x-1"
              style="margin-left: 1rem;"
            >
              <button
                class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
              >
                <svg
                  class="fill-white"
                />
              </button>
              <pre
                class="text-amber-700"
              >
                barbes
                :
              </pre>
               
              <div
                class="flex-auto"
              >
                <pre
                  class="whitespace-pre-wrap break-all text-sky-400"
                >
                  "rochechouart"
                </pre>
              </div>
            </div>
            <button
              class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
              style="margin-left: 2rem; width: calc(100% - 2rem); z-index: 2;"
            >
              <div
                class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
              />
              <div
                class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
              >
                <svg
                  class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                />
              </div>
            </button>
            <div
              class="mt-1 flex items-center space-x-1"
              style="margin-left: 1rem;"
            >
              <button
                class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
              >
                <svg
                  class="fill-white"
                />
              </button>
              <pre
                class="text-amber-700"
              >
                version
                :
              </pre>
               
              <div
                class="flex-auto"
              >
                <pre
                  class="whitespace-pre-wrap break-all text-rose-500"
                >
                  1
                </pre>
              </div>
            </div>
            <button
              class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
              style="margin-left: 2rem; width: calc(100% - 2rem); z-index: 2;"
            >
              <div
                class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
              />
              <div
                class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
              >
                <svg
                  class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                />
              </div>
            </button>
            <div
              class="mt-1 flex items-center space-x-1"
              style="margin-left: 1rem;"
            >
              <button
                class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
              >
                <svg
                  class="fill-white"
                />
              </button>
              <pre
                class="text-amber-700"
              >
                toto
                :
              </pre>
               
              <div
                class="flex-auto"
              >
                <pre
                  class="whitespace-pre-wrap break-all text-purple-500"
                >
                  null
                </pre>
              </div>
            </div>
            <button
              class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
              style="margin-left: 2rem; width: calc(100% - 2rem); z-index: 2;"
            >
              <div
                class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
              />
              <div
                class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
              >
                <svg
                  class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                />
              </div>
            </button>
            <div
              class="mt-1 flex items-center space-x-1"
              style="margin-left: 1rem;"
            >
              <button
                class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700 invisible"
              >
                <svg
                  class="fill-white"
                />
              </button>
              <pre
                class="text-amber-700"
              >
                arr
                :
              </pre>
               
              <pre
                class="flex items-center space-x-0.5"
              >
                [
                <button
                  class="mx-0.5 rounded-sm bg-slate-400 hover:bg-slate-500 active:bg-slate-600"
                >
                  <svg />
                </button>
                ]
              </pre>
            </div>
            <div
              class="relative hidden"
            >
              <button
                class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
                style="margin-left: 3rem; width: calc(100% - 3rem); z-index: 3;"
              >
                <div
                  class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
                />
                <div
                  class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
                >
                  <svg
                    class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                  />
                </div>
              </button>
              <div
                class="mt-1 flex items-center space-x-1"
                style="margin-left: 2rem;"
              >
                <button
                  class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
                >
                  <svg
                    class="fill-white"
                  />
                </button>
                <pre
                  class="text-amber-700"
                >
                  0
                  :
                </pre>
                 
                <div
                  class="flex-auto"
                >
                  <pre
                    class="whitespace-pre-wrap break-all text-rose-500"
                  >
                    0.7309955249202127
                  </pre>
                </div>
              </div>
              <button
                class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
                style="margin-left: 3rem; width: calc(100% - 3rem); z-index: 3;"
              >
                <div
                  class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
                />
                <div
                  class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
                >
                  <svg
                    class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                  />
                </div>
              </button>
              <div
                class="mt-1 flex items-center space-x-1"
                style="margin-left: 2rem;"
              >
                <button
                  class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
                >
                  <svg
                    class="fill-white"
                  />
                </button>
                <pre
                  class="text-amber-700"
                >
                  1
                  :
                </pre>
                 
                <div
                  class="flex-auto"
                >
                  <pre
                    class="whitespace-pre-wrap break-all text-rose-500"
                  >
                    0.45152102178420384
                  </pre>
                </div>
              </div>
              <button
                class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
                style="margin-left: 3rem; width: calc(100% - 3rem); z-index: 3;"
              >
                <div
                  class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
                />
                <div
                  class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
                >
                  <svg
                    class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                  />
                </div>
              </button>
              <div
                class="mt-1 flex items-center space-x-1"
                style="margin-left: 2rem;"
              >
                <button
                  class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
                >
                  <svg
                    class="fill-white"
                  />
                </button>
                <pre
                  class="text-amber-700"
                >
                  2
                  :
                </pre>
                 
                <div
                  class="flex-auto"
                >
                  <pre
                    class="whitespace-pre-wrap break-all text-rose-500"
                  >
                    0.3101491368469109
                  </pre>
                </div>
              </div>
              <button
                class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
                style="margin-left: 3rem; width: calc(100% - 3rem); z-index: 3;"
              >
                <div
                  class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
                />
                <div
                  class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
                >
                  <svg
                    class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                  />
                </div>
              </button>
              <div
                class="mt-1 flex items-center space-x-1"
                style="margin-left: 2rem;"
              >
                <button
                  class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
                >
                  <svg
                    class="fill-white"
                  />
                </button>
                <pre
                  class="text-amber-700"
                >
                  3
                  :
                </pre>
                 
                <div
                  class="flex-auto"
                >
                  <pre
                    class="whitespace-pre-wrap break-all text-rose-500"
                  >
                    0.33964078594296154
                  </pre>
                </div>
              </div>
              <button
                class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
                style="margin-left: 3rem; width: calc(100% - 3rem); z-index: 3;"
              >
                <div
                  class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
                />
                <div
                  class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
                >
                  <svg
                    class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                  />
                </div>
              </button>
              <div
                class="mt-1 flex items-center space-x-1"
                style="margin-left: 2rem;"
              >
                <button
                  class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
                >
                  <svg
                    class="fill-white"
                  />
                </button>
                <pre
                  class="text-amber-700"
                >
                  4
                  :
                </pre>
                 
                <div
                  class="flex-auto"
                >
                  <pre
                    class="whitespace-pre-wrap break-all text-rose-500"
                  >
                    0.579201406715314
                  </pre>
                </div>
              </div>
              <button
                class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
                style="margin-left: 3rem; width: calc(100% - 3rem); z-index: 3;"
              >
                <div
                  class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
                />
                <div
                  class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
                >
                  <svg
                    class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                  />
                </div>
              </button>
            </div>
            <button
              class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
              style="margin-left: 2rem; width: calc(100% - 2rem); z-index: 2;"
            >
              <div
                class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
              />
              <div
                class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
              >
                <svg
                  class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                />
              </div>
            </button>
            <div
              class="mt-1 flex items-center space-x-1"
              style="margin-left: 1rem;"
            >
              <button
                class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700 invisible"
              >
                <svg
                  class="fill-white"
                />
              </button>
              <pre
                class="text-amber-700"
              >
                nested
                :
              </pre>
               
              <pre
                class="flex items-center space-x-0.5"
              >
                {
                <button
                  class="mx-0.5 rounded-sm bg-slate-400 hover:bg-slate-500 active:bg-slate-600"
                >
                  <svg />
                </button>
                }
              </pre>
            </div>
            <div
              class="relative hidden"
            >
              <button
                class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
                style="margin-left: 3rem; width: calc(100% - 3rem); z-index: 3;"
              >
                <div
                  class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
                />
                <div
                  class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
                >
                  <svg
                    class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                  />
                </div>
              </button>
              <div
                class="mt-1 flex items-center space-x-1"
                style="margin-left: 2rem;"
              >
                <button
                  class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
                >
                  <svg
                    class="fill-white"
                  />
                </button>
                <pre
                  class="text-amber-700"
                >
                  thx
                  :
                </pre>
                 
                <div
                  class="flex-auto"
                >
                  <pre
                    class="whitespace-pre-wrap break-all text-rose-500"
                  >
                    1138
                  </pre>
                </div>
              </div>
              <button
                class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
                style="margin-left: 3rem; width: calc(100% - 3rem); z-index: 3;"
              >
                <div
                  class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
                />
                <div
                  class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
                >
                  <svg
                    class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                  />
                </div>
              </button>
              <div
                class="mt-1 flex items-center space-x-1"
                style="margin-left: 2rem;"
              >
                <button
                  class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
                >
                  <svg
                    class="fill-white"
                  />
                </button>
                <pre
                  class="text-amber-700"
                >
                  star
                  :
                </pre>
                 
                <div
                  class="flex-auto"
                >
                  <pre
                    class="whitespace-pre-wrap break-all text-sky-400"
                  >
                    "wars"
                  </pre>
                </div>
              </div>
              <button
                class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
                style="margin-left: 3rem; width: calc(100% - 3rem); z-index: 3;"
              >
                <div
                  class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
                />
                <div
                  class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
                >
                  <svg
                    class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                  />
                </div>
              </button>
            </div>
            <button
              class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
              style="margin-left: 2rem; width: calc(100% - 2rem); z-index: 2;"
            >
              <div
                class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
              />
              <div
                class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
              >
                <svg
                  class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
                />
              </div>
            </button>
          </div>
          <button
            class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
            style="margin-left: 1rem; width: calc(100% - 1rem); z-index: 1;"
          >
            <div
              class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
            />
            <div
              class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
            >
              <svg
                class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
              />
            </div>
          </button>
          <div
            class="mt-1 flex items-center space-x-1"
            style="margin-left: 0rem;"
          >
            <button
              class="rounded-sm bg-red-500 hover:bg-red-600 active:bg-red-700  invisible"
            >
              <svg
                class="fill-white"
              />
            </button>
            <pre
              class="text-amber-700"
            >
              done
              :
            </pre>
             
            <div
              class="flex-auto"
            >
              <pre
                class="whitespace-pre-wrap break-all text-green-500"
              >
                true
              </pre>
            </div>
          </div>
          <button
            class="absolute h-2 -mt-0.5 first:-mt-1.5 group opacity-0 hover:opacity-100"
            style="margin-left: 1rem; width: calc(100% - 1rem); z-index: 1;"
          >
            <div
              class="absolute left-0 right-0 h-0.5 -translate-y-1/2 bg-slate-300 group-hover:bg-slate-400 group-active:bg-slate-500"
            />
            <div
              class="absolute left-1/2 -translate-x-1/2 -translate-y-1/2 border rounded-sm bg-white border-slate-300 group-hover:border-slate-400 group-active:border-slate-500"
            >
              <svg
                class=" fill-slate-300  group-hover:fill-slate-400 group-active:fill-slate-500"
              />
            </div>
          </button>
        </div>
      </div>
    </div>
  </main>
</div>
`;
